redsky/streamer.py

```diff
@@ -13,60 +13,325 @@
+        The main way that this works is by a) ingesting documents, b) issuing
+        documents, c) returning the internals of documents upon request.
+
+        Parameters
+        ----------
+        input_info: list of tuples
+            describs the incoming streams
+        output_info: list of tuples
+            describs the resulting stream
+        provenance : dict, optional
+            metadata about this operation
+
+        Notes
+        ------
+        input_info is designed to map keys in streams to kwargs in functions.
+        It is critical for the internal data from the events to be returned,
+        upon `event_guts`.
+        input_info = [('input_kwarg', 'data_key')]
+
+        output_info is designed to take the output tuple and map it back into
+        data_keys.
+        output_info = [('data_key', {'dtype': 'array', 'source': 'testing'})]
+        """
+        if output_info is None:
+            output_info = {}
+        if input_info is None:
+            input_info = {}
+
+        self.event_failed = False
+        self.run_start_uid = None
+        self.input_info = input_info
+        self.output_info = output_info
+        self.i = None
+        self.outbound_descriptor_uid = None
+        self.provenance = {}
+
+    def generate_provenance(self, func=None):
+        d = dict(
+            stream_class=self.__class__.__name__,
+            stream_class_module=self.__class__.__module__,
+            # TODO: Need to support pip and other sources at some point
+            # conda_list=subprocess.check_output(['conda', 'list',
+            #                                     '-e']).decode()
+        )
+        if self.input_info:
+            d.update(input_info=self.input_info)
+        if self.output_info:
+            d.update(output_info=self.output_info)
+        if func:
+            d.update(function_module=func.__module__,
+                     # this line gets more complex with classes
+                     function_name=func.__name__, )
+        self.provenance = d
+
+    def curate_streams(self, nds):
+        # If we get multiple streams make (doc, doc, doc, ...)
+        if isinstance(nds[0], tuple):
+            names, docs = list(zip(*nds))
+            if len(set(names)) > 1:
+                raise RuntimeError('Misaligned Streams')
+            name = names[0]
+        # If only one stream then (doc, )
+        else:
+            names, docs = nds
+            name = names
+            docs = (docs,)
+        return name, docs
+
+    def dispatch(self, nds):
+        """
+        Dispatch to methods expecting particular doc types.
+
+        Notes
+        ------
+        This takes in a tuple ((name, document), (name, document)).
+        This then re-packages this as (document, document) and dispatches
+        to the corresponding method.
+        """
+        name, docs = self.curate_streams(nds)
+        return getattr(self, name)(docs)
```

Comment:
oh ok found it line 21 of callbacks/core.py in current branch. I would be tempted to point that out to them.